### PR TITLE
Mesh reader/writer with ADIOS2

### PR DESCRIPTION
**Pull request by jorgensd**
This is the first step towards having function check-pointing in DOLFINx.

The code uses `ADIOS2` as backend for reading and writing data.
No MPI communication required in either stage (until mesh creation/partitioning).

The following C++ code runs nicely
```c++

#include <dolfinx/common/MPI.h>
#include <dolfinx/io/ADIOS2Writers.h>
#include <dolfinx/mesh/Mesh.h>
#include <dolfinx/mesh/generation.h>
#include <filesystem>
#include <mpi.h>

using namespace dolfinx;

int main(int argc, char* argv[])
{
  dolfinx::init_logging(argc, argv);
  MPI_Init(&argc, &argv);

  {
    auto mesh
        = std::make_shared<mesh::Mesh<float>>(mesh::create_rectangle<float>(
            MPI_COMM_WORLD, {{{0.0, 0.0}, {1.0, 1.0}}}, {2, 3},
            mesh::CellType::triangle,
            mesh::create_cell_partitioner(mesh::GhostMode::none)));

#ifdef HAS_ADIOS2
    std::filesystem::__cxx11::path filename = "test";
    io::Checkpointer<float> outfile(mesh->comm(),
                                    filename.replace_extension("bp"), mesh);
    outfile.write(2.0);
    outfile.close();
    io::Checkpointer<float> infile(mesh->comm(),
                                   filename.replace_extension("bp"));
    mesh::Mesh<float> mesh_no_ghost
        = infile.read_mesh(2.0, mesh::GhostMode::none);
    mesh::Mesh<float> mesh_shared
        = infile.read_mesh(2.0, mesh::GhostMode::shared_facet);

    std::stringstream cc;
    const int rank = dolfinx::MPI::rank(MPI_COMM_WORLD);
    cc << "-" << rank << "-"
       << "\n";
    cc << "Mesh no ghosts: "
       << mesh_no_ghost.topology()
              ->index_map(mesh_no_ghost.topology()->dim())
              ->size_local()
       << " "
       << mesh_no_ghost.topology()
              ->index_map(mesh_no_ghost.topology()->dim())
              ->num_ghosts()
       << "\n";
    cc << "Mesh shared_facet: "
       << mesh_shared.topology()
              ->index_map(mesh_shared.topology()->dim())
              ->size_local()
       << " "
       << mesh_shared.topology()
              ->index_map(mesh_shared.topology()->dim())
              ->num_ghosts()
       << "\n";
    infile.close();

    io::Checkpointer<float> infile_self(MPI_COMM_SELF,
                                        filename.replace_extension("bp"));
    mesh::Mesh<float> mesh_self
        = infile_self.read_mesh(2.0, mesh::GhostMode::none);
    infile_self.close();
    cc << "MEsh self: "
       << mesh_self.topology()
              ->index_map(mesh_self.topology()->dim())
              ->size_local()
       << " "
       << mesh_self.topology()
              ->index_map(mesh_self.topology()->dim())
              ->num_ghosts()
       << "\n";
    std::cout << cc.str();

    io::VTXWriter<float> vtx_ng(
        mesh_no_ghost.comm(), "mesh_no_ghost.bp",
        std::make_shared<mesh::Mesh<float>>(mesh_no_ghost));
    vtx_ng.write(0.0);
    vtx_ng.close();

    io::VTXWriter<float> vtx_s(
        mesh_no_ghost.comm(), "mesh_shared.bp",
        std::make_shared<mesh::Mesh<float>>(mesh_shared));
    vtx_s.write(0.0);
    vtx_s.close();

    io::VTXWriter<float> vtx_self(
        mesh_self.comm(), "mesh" + std::to_string(rank) + ".bp",
        std::make_shared<mesh::Mesh<float>>(mesh_self));
    vtx_self.write(0.0);
    vtx_self.close();

#endif
  }
  MPI_Finalize();

  return 0;
}

```

```bash
 mpirun -n 3 ./build-dir-real/demo_interpolation-io 
-0-
Mesh no ghosts: 4 0
Mesh shared_facet: 4 2
MEsh self: 12 0
-1-
Mesh no ghosts: 4 0
Mesh shared_facet: 4 2
MEsh self: 12 0
-2-
Mesh no ghosts: 4 0
Mesh shared_facet: 4 4
MEsh self: 12 0
```
It would be good to get feedback on this design prior to implementing the python interface.

Once this is merged, I can start working on the function checkpointing, with the approach used in https://github.com/jorgensd/adios4dolfinx/blob/main/src/adios4dolfinx/checkpointing.py#L91-L173

### CODE REVIEW

File: cpp/dolfinx/io/ADIOS2Writers.h [hunk 1]
```diff
@@ -867,6 +867,330 @@ void vtx_write_mesh_from_space(adios2::IO& io, adios2::Engine& engine,
...
+/// @param[in] mesh The mesh
+template <typename T>
+void write_mesh(adios2::IO& io, adios2::Engine& engine,
+                const mesh::Mesh<T>& mesh)
+{
+  const mesh::Geometry<T>& geometry = mesh.geometry();
+
+  // "Put" geometry
+  std::int32_t num_xdofs_local = geometry.index_map()->size_local();
+  std::int32_t num_xdofs_global = geometry.index_map()->size_global();
```
Comment (garth-wells) [inline cpp/dolfinx/io/ADIOS2Writers.h:885]: Needs to be 64-bit.

File: cpp/dolfinx/io/ADIOS2Writers.h [hunk 2]
```diff
@@ -867,6 +867,330 @@ void vtx_write_mesh_from_space(adios2::IO& io, adios2::Engine& engine,
...
+  std::array<std::int64_t, 2> local_range = geometry.index_map()->local_range();
+
+  std::span<const T> nodes = geometry.x();
```
Comment (garth-wells) [inline cpp/dolfinx/io/ADIOS2Writers.h:888]: We normally just use `x`.

File: cpp/dolfinx/io/ADIOS2Writers.h [hunk 3]
```diff
@@ -867,6 +867,330 @@ void vtx_write_mesh_from_space(adios2::IO& io, adios2::Engine& engine,
...
+  std::size_t gdim = geometry.dim();
+  std::size_t num_nodes = nodes.size() / 3;
+  // Truncate geometry to gdim
+  std::vector<T> geometry_trunc;
```
Comment (garth-wells) [inline cpp/dolfinx/io/ADIOS2Writers.h:892]: Wrap with `mdspan` for readability.

File: cpp/dolfinx/io/ADIOS2Writers.h [hunk 4]
```diff
@@ -867,6 +867,330 @@ void vtx_write_mesh_from_space(adios2::IO& io, adios2::Engine& engine,
 }
 } // namespace impl_vtx
 
+/// @privatesection
+namespace impl_checkpoint
+{
+/// Write mesh to file using VTX format
```
Comment (garth-wells) [inline cpp/dolfinx/io/ADIOS2Writers.h:873]: How is this different from the current VTX mesh writer?

File: cpp/dolfinx/io/ADIOS2Writers.h [hunk 5]
```diff
@@ -867,6 +867,330 @@ void vtx_write_mesh_from_space(adios2::IO& io, adios2::Engine& engine,
...
+/// @param[in] io The ADIOS2 io object
+/// @param[in] engine The ADIOS2 engine object
+/// @param[in] mesh The mesh
+template <typename T>
+void write_mesh(adios2::IO& io, adios2::Engine& engine,
```
Comment (garth-wells) [inline cpp/dolfinx/io/ADIOS2Writers.h:878]: Use more descriptive name if this writes VTX.

File: cpp/dolfinx/io/ADIOS2Writers.h [hunk 6]
```diff
@@ -867,6 +867,330 @@ void vtx_write_mesh_from_space(adios2::IO& io, adios2::Engine& engine,
...
+    // If we have no functions write the mesh to file
+    // if (this->_u.empty())
+
+    this->_engine->EndStep();
+  }
+
+  /// @brief  Read mesh from a given time-step
+  /// @param[in] t The time step
+  /// @param[in] mode The ghost mode
+  mesh::Mesh<T> read_mesh(double t, mesh::GhostMode mode)
```
Comment (garth-wells) [inline cpp/dolfinx/io/ADIOS2Writers.h:1041]: What format does it read?

**Review by garth-wells — CHANGES_REQUESTED** [review]
Some preliminary comments.

Could this be divided into a 'read' PR, and a checkpoint PR after?

**Comment by jorgensd** [comment]
> Some preliminary comments.
> 
> Could this be divided into a 'read' PR, and a checkpoint PR after?

No, as this reads the mesh from the mesh format this PR creates a writer for

File: cpp/dolfinx/io/ADIOS2Writers.h (hunk 5 above)
Comment (jorgensd) [reply cpp/dolfinx/io/ADIOS2Writers.h:878]: It does not write VTX. It writes a pure binary file (that cannot be used for visualization, similar to HDF5File in legacy dolfin).

File: cpp/dolfinx/io/ADIOS2Writers.h (hunk 6 above)
Comment (jorgensd) [reply cpp/dolfinx/io/ADIOS2Writers.h:1041]: `.bp` files created with the `ADIOS2Checkpointer`

File: cpp/dolfinx/io/ADIOS2Writers.h (hunk 5 above)
Comment (garth-wells) [reply cpp/dolfinx/io/ADIOS2Writers.h:878]: > It does not write VTX. It writes a pure binary file (that cannot be used for visualization, similar to HDF5File in legacy dolfin).

So the docstring needs updating? Currently says `/// Write mesh to file using VTX format`.

File: cpp/dolfinx/io/ADIOS2Writers.h (hunk 4 above)
Comment (jorgensd) [reply cpp/dolfinx/io/ADIOS2Writers.h:873]: VTX mesh writer writes local partitions of the mesh (local indices) to separate state arrays in a .bp file.
This means that it cannot be read in on any other number of processes than the number that have written them to file (making them useless for checkpointing).

This mesh writer uses global mesh indices and writes everything to a global array.

**Comment by jorgensd** [comment]
@garth-wells  Any further comments?

**Comment by garth-wells** [comment]
@jorgensd could you make clearer in the PR comment what this changes does in practice?

File: cpp/dolfinx/io/ADIOS2Writers.h [hunk 7]
```diff
@@ -867,6 +867,333 @@ void vtx_write_mesh_from_space(adios2::IO& io, adios2::Engine& engine,
...
+    for (std::size_t j = 0; j < gdim; ++j)
+      geometry_md(i, j) = x[3 * i + j];
+
+  // Define geometry variable and "put"
+  adios2::Variable<T> local_geometry = impl_adios2::define_variable<T>(
+      io, "geometry", {num_xdofs_global, gdim}, {local_range[0], 0},
+      {num_xdofs_local, gdim});
+  engine.Put<T>(local_geometry, geometry_b.data());
+
+  // Put topology (connectivity)
```
Comment (garth-wells) [inline cpp/dolfinx/io/ADIOS2Writers.h:906]: This doesn't 'put', it just access the topology. Why the mutable version?

File: cpp/dolfinx/io/ADIOS2Writers.h [hunk 8]
```diff
@@ -867,6 +867,333 @@ void vtx_write_mesh_from_space(adios2::IO& io, adios2::Engine& engine,
...
+
+  std::span<const std::int32_t> local_dmap(geometry_dmap.data_handle(),
+                                           num_cells_local * num_dofs_per_cell);
+  geometry_imap->local_to_global(local_dmap, connectivity_out);
+
+  // Put local topology in global dataset
+  adios2::Variable<std::int64_t> local_topology
+      = impl_adios2::define_variable<std::int64_t>(
+          io, "connectivity",
+          {(std::size_t)num_cells_global, (std::size_t)num_dofs_per_cell},
```
Comment (garth-wells) [inline cpp/dolfinx/io/ADIOS2Writers.h:937]: Avoid casts by using the required type earlier when assigning.

File: cpp/dolfinx/io/ADIOS2Writers.h [hunk 9]
```diff
@@ -867,6 +867,333 @@ void vtx_write_mesh_from_space(adios2::IO& io, adios2::Engine& engine,
...
+          {(std::size_t)local_range_cell[0], (std::size_t)0},
+          {(std::size_t)num_cells_local, (std::size_t)num_dofs_per_cell});
+  engine.Put<std::int64_t>(local_topology, connectivity_out.data());
+
+  // Compute mesh permutations
```
Comment (garth-wells) [inline cpp/dolfinx/io/ADIOS2Writers.h:942]: Comment doesn't really describe what's happening below.

File: cpp/dolfinx/io/ADIOS2Writers.h [hunk 10]
```diff
@@ -867,6 +867,333 @@ void vtx_write_mesh_from_space(adios2::IO& io, adios2::Engine& engine,
...
+          io, "CellPermutations", {(std::size_t)num_cells_global},
+          {(std::size_t)local_range_cell[0]}, {(std::size_t)num_cells_local});
+  engine.Put<std::uint32_t>(local_cell_perm, cell_perm.data());
+
+  engine.PerformPuts();
+}
+
+} // namespace impl_checkpoint
+
+template <std::floating_point T>
```
Comment (garth-wells) [inline cpp/dolfinx/io/ADIOS2Writers.h:957]: Add docstring.

File: cpp/dolfinx/io/ADIOS2Writers.h [hunk 11]
```diff
@@ -867,6 +867,333 @@ void vtx_write_mesh_from_space(adios2::IO& io, adios2::Engine& engine,
...
+class ADIOS2Checkpointer
```
Comment (garth-wells) [inline cpp/dolfinx/io/ADIOS2Writers.h:958]: In all likelihood, we'll not use anything other than ADIOS2 for checkpointing, so probably don't need ADIOS2 in the name.

Also, code only seems to handle meshes so name is a bit misleading.

File: cpp/dolfinx/io/ADIOS2Writers.h [hunk 12]
```diff
@@ -867,6 +867,333 @@ void vtx_write_mesh_from_space(adios2::IO& io, adios2::Engine& engine,
...
+  ADIOS2Checkpointer(MPI_Comm comm, const std::filesystem::path& filename,
+                     std::shared_ptr<const mesh::Mesh<T>> mesh)
+      : _adios(std::make_unique<adios2::ADIOS>(comm)),
+        _io(std::make_unique<adios2::IO>(
+            _adios->DeclareIO("ADIOS2 checkpoint writer"))),
+        _engine(std::make_unique<adios2::Engine>(
+            _io->Open(filename, adios2::Mode::Write))),
+        _mesh(mesh), _u({}), _comm(comm)
+  {
+    // Define time independent attributes
```
Comment (garth-wells) [inline cpp/dolfinx/io/ADIOS2Writers.h:987]: 'time independent attributes' -> 'topology and geometry'

File: cpp/dolfinx/io/ADIOS2Writers.h [hunk 13]
```diff
@@ -867,6 +867,333 @@ void vtx_write_mesh_from_space(adios2::IO& io, adios2::Engine& engine,
...
+    const mesh::Geometry<T>& geometry = mesh->geometry();
+    if (geometry.cmaps().size() > 1)
+      throw std::runtime_error("Multiple coordinate maps not supported");
+    impl_adios2::define_attribute<int>(*this->_io, "Degree",
+                                       geometry.cmaps()[0].degree());
+    impl_adios2::define_attribute<int>(*this->_io, "LagrangeVariant",
+                                       int(geometry.cmaps()[0].variant()));
+  }
+
+  /// @brief Create a ADIOS2 checkpointer for reading a mesh from file
```
Comment (garth-wells) [inline cpp/dolfinx/io/ADIOS2Writers.h:1004]: Does this need to be a class? Would a function suffice?

File: cpp/dolfinx/io/ADIOS2Writers.h [hunk 14]
```diff
@@ -867,6 +867,333 @@ void vtx_write_mesh_from_space(adios2::IO& io, adios2::Engine& engine,
...
+        = this->_io->template InquireVariable<std::int64_t>("connectivity");
+    if (!var_top)
+    {
+      throw std::runtime_error(
+          "Could not find 'connectivity' variable in input file");
+    }
+    var_top.SetStepSelection(read_selection);
+    adios2::Dims t_shape = var_top.Shape();
+    const std::array<std::int64_t, 2> top_range
+        = MPI::local_range(rank, t_shape[0], size);
```
Comment (garth-wells) [inline cpp/dolfinx/io/ADIOS2Writers.h:1141]: `doflinx::MPI::...` to avoid namespace clashes.

File: cpp/dolfinx/io/ADIOS2Writers.h [hunk 15]
```diff
@@ -867,6 +867,333 @@ void vtx_write_mesh_from_space(adios2::IO& io, adios2::Engine& engine,
...
+    const std::array<std::int64_t, 2> geom_range
+        = MPI::local_range(rank, x_shape[0], size);
+    var_geom.SetSelection({{(std::size_t)geom_range[0], (std::size_t)0},
+                           {std::size_t(geom_range[1] - geom_range[0]),
+                            (std::size_t)x_shape[1]}});
+    std::vector<T> nodes((geom_range[1] - geom_range[0]) * x_shape[1]);
+    this->_engine->Get(var_geom, nodes);
+
+    // Get mesh topology
+    adios2::Variable<std::int64_t> var_top
```
Comment (garth-wells) [inline cpp/dolfinx/io/ADIOS2Writers.h:1131]: I'm not finding the variable names very intuitive.

**Review by garth-wells — CHANGES_REQUESTED** [review]
Another round of comments.

Would be helpful to have clear statement in the PR of what functionality is being add.

**Comment by jorgensd** [comment]
> Another round of comments.
> 
> 
> 
> Would be helpful to have clear statement in the PR of what functionality is being add.

The code can write meshes to file on N processes, and read it back in on M processes. This is the first step towards having full checkpointing, as a next PR would add in writing/reading functions.

This is the reason for making the checkpointer a class, as one would write the mesh and function to the same file (and keep it alive during the process), avoiding duplicate initialization statements.

What differs in this code from the already existing Fides/VTXWriter is that:
1. the data is written to global arrays (so that they can be read in again on a different number of processes).
2. the topology is stored in DOLFINx order, i.e. the dofs are not permuted to VTK or GMSH ordering.

**Comment by garth-wells** [comment]
> What differs in this code from the already existing Fides/VTXWriter is that:
> 
> 1. the data is written to global arrays (so that they can be read in again on a different number of processes).

What's a 'global array'? An ADIOS2 concept?

The VTX files we write do have a local-to-global map array. 

> 2. the topology is stored in DOLFINx order, i.e. the dofs are not permuted to VTK or GMSH ordering.

Do we want to guarantee that the DOLFINx remains unchanged? An advantage of VTK ordering is that we can freely change the internal DOLFINx ordering.

**Comment by jorgensd** [comment]
> What's a 'global array'? An ADIOS2 concept?

Yes. Consider writing out a mesh with the VTXWriter in serial:
```bash
root@dokken-XPS-9320:~/shared/dolfinx/cpp/demo/poisson# ./build-dir-real64/demo_poisson 
root@dokken-XPS-9320:~/shared/dolfinx/cpp/demo/poisson# bpls -l mesh.bp/
  uint32_t  NumberOfCells        {1} = 1024 / 1024
  uint32_t  NumberOfNodes        {1} = 561 / 561
  int64_t   connectivity         [1]*{1024, 4} = 0 / 560
  double    geometry             [1]*{561, 3} = 0 / 2
  double    step                 scalar = 0
  uint32_t  types                scalar = 69
  uint8_t   vtkGhostType         [1]*{561} = 0 / 0
  int64_t   vtkOriginalPointIds  [1]*{561} = 0 / 560
```
is what you get in serial, but in parallel you get
```bash
root@dokken-XPS-9320:~/shared/dolfinx/cpp/demo/poisson# mpirun -n 4 ./build-dir-real64/demo_poisson 
root@dokken-XPS-9320:~/shared/dolfinx/cpp/demo/poisson# bpls -l mesh.bp/
  uint32_t  NumberOfCells        {4} = 272 / 288
  uint32_t  NumberOfNodes        {4} = 166 / 182
  int64_t   connectivity         [4]*{__, 4} = 0 / 181
  double    geometry             [4]*{__, 3} = 0 / 2
  double    step                 scalar = 0
  uint32_t  types                scalar = 69
```
which means that the data is split into block structures (equal to the number of processors). 

This makes the data very fast to write, but very hard to read on any other number of processes than the one that wrote it.

A global array is what we are creating when we specify both local and global dimensions in `put` as well as the global offset, like:
```c++
adios2::Variable<T> local_geometry = impl_adios2::define_variable<T>(
      io, "geometry", {num_xdofs_global, gdim}, {local_range[0], 0},
      {num_xdofs_local, gdim});
```
which gives you a `.bp` file that looks like:
```bash
root@dokken-XPS-9320:~/shared/dolfinx/cpp/demo/checkpoint# mpirun -n 4 ./build-dir-real64/demo_interpolation-io 
root@dokken-XPS-9320:~/shared/dolfinx/cpp/demo/checkpoint# bpls -l test.bp/
  uint32_t  CellPermutations  {12} = 0 / 6
  int64_t   connectivity      {12, 3} = 0 / 11
  float     geometry          {12, 2} = 0 / 1
  double    step              scalar = 2
```

File: cpp/dolfinx/io/ADIOS2Writers.h (hunk 12 above)
Comment (jorgensd) [reply cpp/dolfinx/io/ADIOS2Writers.h:952]: Not sure what you mean by this. The topology can't change (but it is not an attribute, it is a variable). The geometry can change over time.

File: cpp/dolfinx/io/ADIOS2Writers.h (hunk 13 above)
Comment (jorgensd) [reply cpp/dolfinx/io/ADIOS2Writers.h:969]: Maybe, it depends on how we want to do it when adding functions to the mix. have stand-alone functions for each of the operations, or having them interact in a class.

File: cpp/dolfinx/io/ADIOS2Writers.h [hunk 16]
```diff
@@ -830,6 +830,337 @@ void vtx_write_mesh_from_space(adios2::IO& io, adios2::Engine& engine,
...
+            _io->Open(filename, adios2::Mode::Write))),
+        _mesh(mesh), _u({}), _comm(comm)
+  {
+    // Define time-independent topology attributes (cell type)
+    auto topology = mesh->topology();
+    assert(topology);
+    if (topology->cell_types().size() > 1)
+      throw std::runtime_error("Multiple cell types not supported");
+    impl_adios2::define_attribute<int>(*this->_io, "CellType",
+                                       int(topology->cell_types()[0]));
```
Comment (jhale) [inline cpp/dolfinx/io/ADIOS2Writers.h:956]: static_cast?

File: cpp/dolfinx/io/ADIOS2Writers.h [hunk 17]
```diff
@@ -830,6 +830,337 @@ void vtx_write_mesh_from_space(adios2::IO& io, adios2::Engine& engine,
...
+    // Get types to reconstruct coordinate map
+    adios2::Attribute<int> ct_var
+        = this->_io->template InquireAttribute<int>("CellType");
+    if (!ct_var)
+    {
+      throw std::runtime_error(
+          "Could not find 'CellType' attribute in input file");
+    }
+
+    mesh::CellType cell_type
```
Comment (jhale) [inline cpp/dolfinx/io/ADIOS2Writers.h:1054]: implied by cast, use auto.

File: cpp/dolfinx/io/ADIOS2Writers.h [hunk 18]
```diff
@@ -830,6 +830,337 @@ void vtx_write_mesh_from_space(adios2::IO& io, adios2::Engine& engine,
...
+          "Could not find 'geometry' variable in input file");
+    }
+    geometry_variable.SetStepSelection(read_selection);
+    adios2::Dims x_shape = geometry_variable.Shape();
+    const std::array<std::int64_t, 2> geom_range
+        = dolfinx::MPI::local_range(rank, x_shape[0], size);
+    geometry_variable.SetSelection(
+        {{(std::size_t)geom_range[0], (std::size_t)0},
+         {std::size_t(geom_range[1] - geom_range[0]),
+          (std::size_t)x_shape[1]}});
```
Comment (jhale) [inline cpp/dolfinx/io/ADIOS2Writers.h:1093]: x_shape already size_t?

File: cpp/dolfinx/io/ADIOS2Writers.h [hunk 19]
```diff
@@ -830,6 +830,337 @@ void vtx_write_mesh_from_space(adios2::IO& io, adios2::Engine& engine,
...
+                                                 std::move(offset));
+    mesh::Mesh<T> mesh
+        = mesh::create_mesh(_comm.comm(), cells_adj, {cmap}, nodes,
+                            {geom_range[1] - geom_range[0], x_shape[1]}, mode);
+
+    return mesh;
+  }
+
+  /// @brief Move constructor
+  Checkpointer(Checkpointer&& writer) = default;
```
Comment (jhale) [inline cpp/dolfinx/io/ADIOS2Writers.h:1131]: noexcept?

File: cpp/dolfinx/io/ADIOS2Writers.h [hunk 20]
```diff
@@ -830,6 +830,337 @@ void vtx_write_mesh_from_space(adios2::IO& io, adios2::Engine& engine,
...
+
+  /// @brief Copy constructor
+  Checkpointer(const Checkpointer&) = delete;
+
+  /// @brief Destructor
+  ~Checkpointer() { close(); }
+
+  /// @brief Move assignment
+  Checkpointer& operator=(Checkpointer&& writer) = default;
```
Comment (jhale) [inline cpp/dolfinx/io/ADIOS2Writers.h:1140]: noexcept?

File: cpp/dolfinx/io/ADIOS2Writers.h [hunk 21]
```diff
@@ -830,6 +830,337 @@ void vtx_write_mesh_from_space(adios2::IO& io, adios2::Engine& engine,
...
+        = static_cast<mesh::CellType>(ct_var.Data().front());
+
+    adios2::Attribute<int> lv_var
+        = this->_io->template InquireAttribute<int>("LagrangeVariant");
+    if (!lv_var)
+    {
+      throw std::runtime_error(
+          "Could not find 'LagrangeVariant' attribute in input file");
+    }
+    basix::element::lagrange_variant lagrange_variant
```
Comment (jhale) [inline cpp/dolfinx/io/ADIOS2Writers.h:1064]: implied by cast use auto

File: cpp/dolfinx/io/ADIOS2Writers.h [hunk 22]
```diff
@@ -830,6 +830,337 @@ void vtx_write_mesh_from_space(adios2::IO& io, adios2::Engine& engine,
...
+    {
+      throw std::runtime_error(
+          "Could not find 'geometry' variable in input file");
+    }
+    geometry_variable.SetStepSelection(read_selection);
+    adios2::Dims x_shape = geometry_variable.Shape();
+    const std::array<std::int64_t, 2> geom_range
+        = dolfinx::MPI::local_range(rank, x_shape[0], size);
+    geometry_variable.SetSelection(
+        {{(std::size_t)geom_range[0], (std::size_t)0},
```
Comment (jhale) [inline cpp/dolfinx/io/ADIOS2Writers.h:1091]: static cast?

File: cpp/dolfinx/io/ADIOS2Writers.h [hunk 23]
```diff
@@ -830,6 +830,333 @@ void vtx_write_mesh_from_space(adios2::IO& io, adios2::Engine& engine,
...
+    // "Put" geometry
+    std::size_t num_xdofs_local = geometry.index_map()->size_local();
+    std::int64_t num_xdofs_global = geometry.index_map()->size_global();
+    std::array<std::int64_t, 2> local_range
+        = geometry.index_map()->local_range();
+
+    std::span<const T> x = geometry.x();
+    std::size_t gdim = geometry.dim();
+
+    // Trim geometry to gdim and trunctate for owned entities only
```
Comment (garth-wells) [inline cpp/dolfinx/io/ADIOS2Writers.h:856]: This is a bit unusual for a 'checkpoint', which I would expect would align closely with the internal storage.

File: cpp/dolfinx/io/ADIOS2Writers.h [hunk 24]
```diff
@@ -830,6 +830,333 @@ void vtx_write_mesh_from_space(adios2::IO& io, adios2::Engine& engine,
...
+  if (geometry.cmaps().size() > 1)
+    throw std::runtime_error("Multiple coordinate maps not supported");
+  fem::ElementDofLayout geom_layout = geometry.cmaps()[0].create_dof_layout();
+  std::size_t num_dofs_per_cell
+      = geom_layout.num_entity_closure_dofs(geometry.dim());
+  auto geometry_imap = geometry.index_map();
+  auto geometry_dmap = geometry.dofmap();
+
+  // Convert local geometry to global geometry
+  std::vector<std::int64_t> connectivity_out(num_cells_local
```
Comment (garth-wells) [inline cpp/dolfinx/io/ADIOS2Writers.h:887]: The terminology here is confusing. The variable name includes 'connectivity', but that data that gets copied in is geometry data.

File: cpp/dolfinx/io/ADIOS2Writers.h [hunk 25]
```diff
@@ -830,6 +830,333 @@ void vtx_write_mesh_from_space(adios2::IO& io, adios2::Engine& engine,
...
+
+  // Put local topology in global dataset
+  adios2::Variable<std::int64_t> local_topology
+      = impl_adios2::define_variable<std::int64_t>(
+          io, "connectivity", {num_cells_global, num_dofs_per_cell},
+          {local_range_cell, 0}, {num_cells_local, num_dofs_per_cell});
+  engine.Put<std::int64_t>(local_topology, connectivity_out.data());
+
+  // Compute mesh permutations and store them to file. These are only
+  // needed for when we want to checkpoint functions.
```
Comment (garth-wells) [inline cpp/dolfinx/io/ADIOS2Writers.h:901]: 'when we want to checkpoint fem::Functions.'

File: cpp/dolfinx/io/ADIOS2Writers.h [hunk 26]
```diff
@@ -830,6 +830,333 @@ void vtx_write_mesh_from_space(adios2::IO& io, adios2::Engine& engine,
...
+  engine.Put<std::uint32_t>(local_cell_perm, cell_perm.data());
+
+  engine.PerformPuts();
+}
+
+} // namespace impl_checkpoint
+
+/// Read and write mesh::Mesh objects with ADIOS2
+template <std::floating_point T>
+class Checkpointer
```
Comment (garth-wells) [inline cpp/dolfinx/io/ADIOS2Writers.h:918]: It doesn't seem logical that this class write and read - beyond the ADIOS2 handles the member data and methods are not shared across read and write. And the `ADIOS2Writer` class seems to already handle the ADISO2 handles.

File: cpp/dolfinx/io/ADIOS2Writers.h [hunk 27]
```diff
@@ -830,6 +830,333 @@ void vtx_write_mesh_from_space(adios2::IO& io, adios2::Engine& engine,
...
+    if (*_engine)
+      _engine->Close();
+  }
+
+private:
+  std::unique_ptr<adios2::ADIOS> _adios;
+  std::unique_ptr<adios2::IO> _io;
+  std::unique_ptr<adios2::Engine> _engine;
+  std::shared_ptr<const mesh::Mesh<T>> _mesh;
+  U _u;
```
Comment (garth-wells) [inline cpp/dolfinx/io/ADIOS2Writers.h:1156]: Is this used?

**Review by garth-wells — CHANGES_REQUESTED** [review]
I think checkpointing needs a good high-level plan and discussion before merging. We need a summary of what we want to be able to do to inform the design. Parts of this PR mirror the VTX and Fides writers, but these designs are probably not ideal for different types.

When reading a file we don't know a priori what the types are. We need to able to (i) query this, or (ii) maybe return a `std::variant`.

**Pull request closed without merging**